### ISSUE TITLE
New rule: avoid non explicit null check

### ISSUE THREAD

**Issue by codec-abc** [DotNetAnalyzers/StyleCopAnalyzers#3469]
Hi,

I am not sure if this the right place to ask but one common bit of frustrating syntax usage in my current code-base is the usage of implicit null check inside a if condition. By that I mean:

````csharp
var obj = GetStuff();
if (obj)
{
   // some code Here
}
````

instead of 

````csharp
var obj = GetStuff();
if (ob != null)
{
    // some code Here
}
````
The second version looks better to me because it is explicit. I read every analyzer documentation in this repo and I didn't find one that does this. So if it the case what do you think of a rule to detect and hopeful fix this?

**Comment by BDisp** [comment]
Isn't depend of what the `GetStuff()` returns? If returns a `bool` then the first looks better. If it returns a `type` then the second is preferable. If it's a `void` method it will return an exception.

**Comment by codec-abc** [comment]
Yeah of course. It only make senses for any type that isn't a boolean. Arguably it could change the behavior for type that have an override for the bool operator. Because you will go in that very case from a implicit bool conversion to a null test. Still, I like languages where an if test block should always evaluate to a boolean and having "behind the scene" null checks is a bad idea imo, thus the reason why I am asking this.

**Comment by BDisp** [comment]
Yes I agree with you. I forgot about `nullable bool`. I also rather null checks on front.

**Comment by sharwell** [comment]
The first case is not valid C# syntax, so I'm not sure what the request here is.

**Comment by codec-abc** [comment]
My bad, I see this code so much often these days that I thought it was legit C# code. I now feel like an idiot for opening such an issue and missing the bigger picture. The thing is that the API I use (Unity's) define a custom bool operator for their base class that return true when the object is non null and false when null (see [here for reference](https://docs.unity3d.com/ScriptReference/Object-operator_Object.html). I find this rather confusing as many variables in the code base I am working with are poorly named and it hard to tell which one is a boolean and which one a Unity's object just by name, which in turn make it hard to deduce the intent of the test. So the initial request might be rephrased as "Avoid using implicit bool operator in if test clause", but I guess it is a little specific to this use case. From a more global POV, one might argue that there could be a warning against using implicit operator at all since it is quite easy to not realize that a function is called when using them.

**Comment by sharwell** [comment]
Yes, for that case I would recommend deleting the implicit conversion to `bool`.

**Comment by codec-abc** [comment]
How so? Is this possible to override it? Because in this case it is defined in Unity API which is closed source and 3rd parties are probably also taking advantage of it.

**Comment by sharwell** [comment]
[BannedApiAnalyzers](https://github.com/dotnet/roslyn-analyzers/blob/main/src/Microsoft.CodeAnalysis.BannedApiAnalyzers/BannedApiAnalyzers.Help.md) should be able to prevent use of this operator. If it doesn't currently support banning implicit operators, that would be a bug we'd like to fix.

**Comment by codec-abc** [comment]
Seems neat. I will give this a shot. Thanks for the help!

**Comment by codec-abc** [comment]
@sharwell 
I am trying to test what you said yesterday by using the analyzer for banned APIs. I struggle a bit with the format for the operator bool. I have this:
````
M:UnityEngine.Object.operator bool(UnityEngine.Engine.Object);Use explicit null check
````
but it doesn't work. Yet the thing is configured correctly since I tested other method and the warning showed up correctly. What' s the syntax for the impilicit bool operator using this?

**Comment by sharwell** [comment]
You can find the syntax for any given item by enabling XML documentation output for the project, and then finding that item in the resulting XML file. I'm not sure what the syntax for this one would be off the top of my head. :)

**Comment by codec-abc** [comment]
Thanks again! Do you have a reference page for enabling documentation output? I am not sure what you are refering to and would like to make sure I use it correctly.
